Source/Core/Core/PowerPC/JitArm64/JitArm64_FloatingPoint.cpp

```diff
@@ -386,3 +390,195 @@ void JitArm64::fctiwzx(UGeckoInstruction inst)
   ASSERT_MSG(DYNA_REC, b == d || single == fpr.IsSingle(b, true),
              "Register allocation turned singles into doubles in the middle of fctiwzx");
 }
+
+// Since the following float conversion functions are used in non-arithmetic PPC float
+// instructions, they must convert floats bitexact and never flush denormals to zero or turn SNaNs
+// into QNaNs. This means we can't just use FCVT/FCVTL/FCVTN.
+
+void JitArm64::ConvertDoubleToSingleLower(size_t guest_reg, ARM64Reg dest_reg, ARM64Reg src_reg)
+{
+  if (js.fpr_is_store_safe[guest_reg])
+  {
+    m_float_emit.FCVT(32, 64, EncodeRegToDouble(dest_reg), EncodeRegToDouble(src_reg));
+    return;
+  }
+
+  FlushCarry();
+
+  const BitSet32 gpr_saved = gpr.GetCallerSavedUsed() & BitSet32(0x4000000F);
+  ABI_PushRegisters(gpr_saved);
+
+  m_float_emit.UMOV(64, ARM64Reg::X0, src_reg, 0);
+  BL(cdts);
+  m_float_emit.INS(32, dest_reg, 0, ARM64Reg::W1);
+
+  ABI_PopRegisters(gpr_saved);
+}
+
+void JitArm64::ConvertDoubleToSinglePair(size_t guest_reg, ARM64Reg dest_reg, ARM64Reg src_reg)
+{
+  if (js.fpr_is_store_safe[guest_reg])
+  {
+    m_float_emit.FCVTN(32, EncodeRegToDouble(dest_reg), EncodeRegToDouble(src_reg));
+    return;
+  }
+
+  FlushCarry();
+
+  const BitSet32 gpr_saved = gpr.GetCallerSavedUsed() & BitSet32(0x4000000F);
+  ABI_PushRegisters(gpr_saved);
+
+  m_float_emit.UMOV(64, ARM64Reg::X0, src_reg, 0);
+  BL(cdts);
+  m_float_emit.INS(32, dest_reg, 0, ARM64Reg::W1);
+
+  m_float_emit.UMOV(64, ARM64Reg::X0, src_reg, 1);
+  BL(cdts);
+  m_float_emit.INS(32, dest_reg, 1, ARM64Reg::W1);
+
+  ABI_PopRegisters(gpr_saved);
+}
+
+void JitArm64::ConvertSingleToDoubleLower(size_t guest_reg, ARM64Reg dest_reg, ARM64Reg src_reg,
+                                          ARM64Reg scratch_reg)
+{
+  ASSERT(scratch_reg != src_reg);
+
+  if (js.fpr_is_store_safe[guest_reg])
+  {
+    m_float_emit.FCVT(64, 32, EncodeRegToDouble(dest_reg), EncodeRegToDouble(src_reg));
+    return;
+  }
+
+  const bool switch_to_farcode = !IsInFarCode();
+
+  FlushCarry();
+
+  // Do we know that the input isn't NaN, and that the input isn't denormal or FPCR.FZ is not set?
```

Comment:
Is this a "TODO:"?

Reply:
No, this is a description of what condition the assembly code below is checking for.

Reply:
Ah, is just that normally question are part of a  TODO. It could read "Check if the input isn't..." and so on

Reply:
Hm, but then I'm not sure how I should phrase the follow-up comments in a way that is clear... (The "if no" and "if yes" comments further down)